x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/FieldSubsetReader.java

```diff
@@ -399,6 +400,20 @@ public void binaryField(FieldInfo fieldInfo, byte[] value) throws IOException {
                 Map<String, Object> transformedSource = filter(result.v2(), filter, 0);
                 XContentBuilder xContentBuilder = XContentBuilder.builder(result.v1().xContent()).map(transformedSource);
                 visitor.binaryField(fieldInfo, BytesReference.toBytes(BytesReference.bytes(xContentBuilder)));
+            } else if (IgnoredSourceFieldMapper.NAME.equals(fieldInfo.name)) {
+                // for _ignored_source, parse, filter out the field and its contents, and serialize back downstream
+                IgnoredSourceFieldMapper.MappedNameValue mappedNameValue = IgnoredSourceFieldMapper.decodeAsMap(value);
+                Map<String, Object> transformedField = filter(mappedNameValue.map(), filter, 0);
+                if (transformedField.isEmpty() == false) {
+                    var topValue = transformedField.values().iterator().next();
+                    if (topValue instanceof Map<?, ?> || topValue instanceof List<?>) {
+                        // The field value contains an object or an array, reconstruct it from filtered values.
+                        visitor.binaryField(fieldInfo, IgnoredSourceFieldMapper.encodeFromMap(mappedNameValue, transformedField));
+                    } else {
+                        // The field value contains no object, so no filtering happened. Use the original value.
```

Comment:
I don't understand the topValue check here. If no filtering happened shouldn't `transformedField` be equal to `mappedNameValue.map()`? I think if a top level field is a leaf field then `topValue` could be not a Map or List? (depending on which value is first pulled from the iterator)

Reply:
The map by construction contains the field name at the top, pointing to the contents. If the contents don't correspond to START_ARRAY or START_OBJECT, this is a leaf field with a single value. If the map is not empty, no filtering has been applied so we can use the incoming value as is.

On the other hand, if the field contains an object or an array, there may be changes deeper in the subobjects and subfields. In this case, we use `transformedField` to play it safe, reconstructing the field value from the (potentially filtered) field contents.

Reply:
I see, but what if the original document (`value` variable here) is like this:

```
field1: value
field2:
  field3: value
```

The only allowed field is field1. 

So `transformedField` would look like this:

```
field1: value
```

Which would cause `value` to be passed down. Which I think is incorrect?

Reply:
My understanding is that this will be treated as two separate calls to this function, as there will be different StoredFields for each. The first one will pass as-is, the second one will be filtered and its`transformedField` will be empty. 

I'll add a yaml test to verify, though the ones with `- '{"name": "A", "object":{ "secret":"mission", "public":"interest" }}'` should be covering this iiuc.

Reply:
Switched to checking the unfiltered map, to play it safe.

Reply:
I see, so if the field in ignored field is an object or an array, then the `mappedNameValue.map()` contains one object / list.  In that  case `transformedField` may not have all the sub keys of the original map (`mappedNameValue.map()`). There for we need to serialize `transformedField`.

Whereas if the field in ignored field is a value and `transformedField` isn't empty, this means it is safe to use the original byte array as an optimization. 

Maybe clarify this a bit in the comment? For example the fact that  each map has only one top level entry, other entries or in different stored fields. And why it is safe to reuse `value` incase the one entry is value.

Reply:
Updated, ptal.